drivers/adc/adc_emul.c

```diff
@@ -0,0 +1,512 @@
+ *
+ * @return 0 on success
+ * @return -ENOTSUP if requested resolution or channel is out side of supported
+ *         range
+ * @return -ENOMEM if buffer is not big enough
+ *         (see @ref adc_emul_validate_buffer_size)
+ * @return other error code returned by adc_context_wait_for_completion
+ */
+static int adc_emul_start_read(const struct device *dev,
+			       const struct adc_sequence *sequence)
+{
+	const struct adc_emul_config *config = dev->config;
+	struct adc_emul_data *data = dev->data;
+	int err;
+
+	if (sequence->resolution > ADC_EMUL_MAX_RESOLUTION ||
+	    sequence->resolution == 0) {
+		LOG_ERR("unsupported resolution %d", sequence->resolution);
+		return -ENOTSUP;
+	}
+
+	if (find_msb_set(sequence->channels) > config->channels_num) {
+		LOG_ERR("unsupported channels in mask: 0x%08x",
+			sequence->channels);
+		return -ENOTSUP;
+	}
+
+	err = adc_emul_validate_buffer_size(dev, sequence);
+	if (err) {
+		LOG_ERR("buffer size too small");
+		return err;
+	}
+
+	data->res_mask = BIT_MASK(sequence->resolution);
+	data->buffer = sequence->buffer;
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_emul_read_async(const struct device *dev,
+			       const struct adc_sequence *sequence,
+			       struct k_poll_signal *async)
+{
+	struct adc_emul_data *data = dev->data;
+	int err;
+
+	adc_context_lock(&data->ctx, async ? true : false, async);
+	err = adc_emul_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+
+static int adc_emul_read(const struct device *dev,
+			 const struct adc_sequence *sequence)
+{
+	return adc_emul_read_async(dev, sequence, NULL);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_emul_data *data = CONTAINER_OF(ctx, struct adc_emul_data,
+						  ctx);
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buffer = data->buffer;
+
+	k_sem_give(&data->sem);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_emul_data *data = CONTAINER_OF(ctx, struct adc_emul_data,
+						  ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+/**
+ * @brief Main function of thread which is used to collect samples from
+ *        emulated ADC. When adc_context_start_sampling give semaphore,
+ *        for each requested channel value function is called. Returned
+ *        mV value is converted to output using reference voltage, gain
+ *        and requested resolution.
+ *
+ * @param data Internal data of ADC emulator
+ *
+ * @return This thread should not end
+ */
+static void adc_emul_acquisition_thread(struct adc_emul_data *data)
+{
+	struct adc_emul_channel_cfg *channel_cfg;
+	adc_emul_res_t result = 0;
+	uint32_t input_mV = 0;
+	uint8_t channel;
+	uint32_t ref_v;
+	uint64_t temp; /* Temporary 64 bit value prevent overflows */
+	int err;
```

Comment:
Can you put some of these declarations inside the loop?

Reply:
Done